sdk/samples/iot/hub/src/paho_iot_hub_pnp_example.c

```diff
@@ -0,0 +1,1020 @@
+static az_span get_request_id(void);
+
+int main(void)
+{
+  int rc;
+
+  // Get the program start time for command response
+  time_t rawtime;
+  struct tm* timeinfo;
+  time(&rawtime);
+  timeinfo = localtime(&rawtime);
+  size_t len = strftime(boot_time_str, sizeof(boot_time_str), iso_spec_time_format, timeinfo);
+  if (len == 0)
+  {
+    printf("Insufficient buffer size for program start time.\n");
+    return -1;
+  }
+  boot_time_span = az_span_init((uint8_t*)boot_time_str, (int32_t)len);
+
+  // Read in the necessary environment variables and initialize the az_iot_hub_client
+  if (az_failed(rc = read_configuration_and_init_client()))
+  {
+    printf("Failed to read configuration from environment variables, return code %d\n", rc);
+    return rc;
+  }
+
+  // Get the MQTT client id used for the MQTT connection
+  size_t client_id_length;
+  if (az_failed(
+          rc = az_iot_hub_client_get_client_id(
+              &client, mqtt_client_id, sizeof(mqtt_client_id), &client_id_length)))
+  {
+    printf("Failed to get MQTT client id, return code %d\n", rc);
+    return rc;
+  }
+
+  // Create the Paho MQTT client
+  if ((rc = MQTTClient_create(
+           &mqtt_client, mqtt_endpoint, mqtt_client_id, MQTTCLIENT_PERSISTENCE_NONE, NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to create MQTT client, return code %d\n", rc);
+    return rc;
+  }
+
+  // Connect to IoT Hub
+  if ((rc = connect_device()) != 0)
+  {
+    return rc;
+  }
+
+  // Subscribe to the necessary twin and commands topics to receive twin updates and responses
+  if ((rc = subscribe()) != 0)
+  {
+    return rc;
+  }
+
+  char* incoming_message_topic;
+  int incoming_message_topic_len;
+  MQTTClient_message* message;
+
+  // First get the twin document to check for updated desired properties. Will then parse desired
+  // property and update accordingly.
+  send_twin_get_message();
+
+  while (device_operational)
+  {
+    // Receive any incoming messages from twin or commands
+    if (MQTTClient_receive(
+            mqtt_client,
+            &incoming_message_topic,
+            &incoming_message_topic_len,
+            &message,
+            TIMEOUT_WAIT_FOR_RECEIVE_MESSAGE_MS)
+            == MQTTCLIENT_SUCCESS
+        && incoming_message_topic != NULL)
+    {
+      on_received(incoming_message_topic, incoming_message_topic_len, message);
+
+      MQTTClient_freeMessage(&message);
+      MQTTClient_free(incoming_message_topic);
+    }
+
+    // Send a telemetry message
+    send_telemetry_message();
+
+    sleep_for_seconds(1);
+  }
+
+  (void)getchar();
+
+  // Gracefully disconnect: send the disconnect packet and close the socket
+  if ((rc = MQTTClient_disconnect(mqtt_client, TIMEOUT_MQTT_DISCONNECT_MS)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to disconnect MQTT client, return code %d\n", rc);
+    return rc;
+  }
+  printf("Disconnected.\n");
```

Comment:
>printf("Disconnected.\n"); [](start = 2, length = 26)

Nit: if the MQTTCLient_disconnect() call fails you may not be disconnected at paho layer (only God knows what state would really be).  I think this printf() should be in an else {...} after the if() check above.

Reply:
good point. changing